docs/source/notebooks/quantum_kernel_methods.ipynb

```diff
@@ -0,0 +1,866 @@
+    "\n",
+    "# Pretty typing\n",
+    "from typing import Callable, Tuple"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "Firstly, we consider an $M$-mode linear interferometer described by the matrix, $U$. We note that the probability of obtaining an output state, $|\\mathbf{t} \\rangle = |t_1, t_2, \\cdots , t_M \\rangle$, for a given input state, $|\\mathbf{s} \\rangle = |s_1, s_2, \\cdots , s_M \\rangle$ is:\n",
+    "$$ |\\langle \\mathbf{s} | U | \\mathbf{t} \\rangle|^{2} = \\frac{|\\text{Per} \\left( U_{s, t} (x, x') \\right)|^2}{t_1! t_2! ... t_M!} $$\n",
+    "\n",
+    "where $\\text{Per}(U_{s, t})$ is the permanent of the matrix, $U_{s, t}$ which is constructed by repeating the $j^{th}$ column $t_{j}$ times, followed by repeating the $i^{th}$ row $s_{i}$ times [2]. First, we will define a function which will estimate the kernel associated with a given feature map, $U(x)$ and initial state, $| s \\rangle.$ \n",
+    "$$ \\kappa^{(\\mathbf{s})}(x, x') = |\\langle \\mathbf{s} | U^{\\dagger}(x') U(x) | \\mathbf{s} \\rangle| ^ {2} = |\\langle \\mathbf{s} | U (x, x') | \\mathbf{s} \\rangle| ^ {2}$$\n",
+    "\n",
+    "We consider two possible kernels with indistinguishable and distinguishable photons. In the case of indistinguishable photons, we have the \"quantum kernel\":\n",
+    "$$ \\kappa^{(\\mathbf{s})}_Q(x, x') = \\frac{|\\text{Per} \\left( U_{s, s} (x, x') \\right)|^2}{\\mathbf{s}!} $$\n",
+    "\n",
+    "where $\\mathbf{s}! = s_1! s_2! \\cdots s_M!$. We will also consider the kernel constructed using distinguishable photons. In this case, we have the classical \"coherent kernel\":\n",
+    "$$ \\kappa_C^{(\\mathbf{s})}(x, x') = \\frac{\\text{Per} \\left( |U_{s, s} (x, x')|^{2} \\right)}{\\mathbf{s}!} $$\n",
+    "\n",
+    "There exists an efficient classical algorithm to calculate the permanent $\\text{Per} \\left( |U_{s, s} (x, x')|^{2} \\right)$ [3].\n",
+    "\n",
+    "In experimental circumstances, the experimentalist may not have access to photon-number resolving (PNR) detectors. In this instance, we consider the \"unbunching kernel\" in which there is maximum 1 photon found in each mode. In this circumstance, we lose the commutative property of the kernel function and the kernel matrix is not symmetric, positive definite for identical input datasets. We will force the kernel matrix to be symmetric by projecting the upper triangular onto the lower triangular. We will also force this matrix to be positive definite by decomposing the matrix spectrally:\n",
+    "$$K = Q \\Lambda Q^{T}$$\n",
+    "and setting the negative eigenvalues in $\\Lambda$ to $0$. The kernel matrix is then reconstructed."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def kernel(\n",
+    "    x1, x2: np.ndarray, \n",
+    "    feature_map: Callable[[np.ndarray], Circuit], \n",
```

Comment:
The typing doesn't match the use of the function since you also have an `adjoint` parameter

Reply:
Hi,
Thanks for all your comments.
Here, `adjoint` is not a parameter of  the `kernel` function. It's a parameter of the specified feature map function. I can simplify the code, however, and invert the feature map within the `kernel` function which would remove `adjoint` as a necessary parameter.